R/surv-roc_survival_auc.R

```diff
@@ -0,0 +1,87 @@
+#' ROC Survival AUC
+#'
+#' Compute the area under the ROC survival curve.
+#'
+#' @family dynamic survival metrics
+#' @templateVar fn roc_auc_survival
+#' @template return-dynamic-survival
+#' @details
+#'
+#' This method automatically groups by the `eval_time` argument.
+#'
+#' Smaller values of the score are associated with better model performance.
+#'
+#' @seealso
+#' Compute the ROC survival curve with [roc_curve_survival()].
+#'
+#' @inheritParams brier_survival
+#'
+#' @param ... Not currently used.
+#'
+#' @author Emil Hvitfeldt
+#'
+#' @examples
+#' library(dplyr)
+#'
+#' lung_surv %>%
+#'   roc_auc_survival(
+#'     truth = surv_obj,
+#'     estimate = .pred_survival,
+#'     censoring_weights = ipcw,
+#'     eval_time = .time
+#'   )
+#' @export
+roc_auc_survival <- function(data, ...) {
+  UseMethod("roc_auc_survival")
+}
+
+roc_auc_survival <- new_dynamic_survival_metric(
+  roc_auc_survival,
+  direction = "minimize"
+)
+
+#' @rdname roc_auc_survival
+#' @export
+roc_auc_survival.data.frame <- function(data,
+                                        truth,
+                                        estimate,
+                                        censoring_weights,
+                                        eval_time,
+                                        na_rm = TRUE,
+                                        case_weights = NULL,
+                                        ...) {
+
+  dynamic_survival_metric_summarizer(
+    name = "roc_auc_survival",
+    fn = roc_auc_survival_vec,
+    data = data,
+    truth = !!enquo(truth),
+    estimate = !!enquo(estimate),
+    censoring_weights = !!enquo(censoring_weights),
+    eval_time = !!enquo(eval_time),
+    na_rm = na_rm,
+    case_weights = !!enquo(case_weights)
+  )
+}
+
+#' @export
+#' @rdname roc_auc_survival
+roc_auc_survival_vec <- function(truth,
+                                 estimate,
+                                 censoring_weights,
+                                 eval_time,
+                                 na_rm = TRUE,
+                                 case_weights = NULL,
+                                 ...) {
+  # No checking since roc_curve_survival_vec() does checking
+  curve <- roc_curve_survival_vec(truth, estimate, censoring_weights, eval_time)
+  roc_trap_auc(curve$specificity, curve$sensitivity)
+}
+
+roc_trap_auc <- function(specificity, sensitivity) {
+  not_na <- !is.na(sensitivity) & !is.na(specificity)
+  sensitivity <- sensitivity[not_na]
+  specificity <- specificity[not_na]
+
+  auc( 1 - specificity, sensitivity)
```

Comment:
Feels worth noting that a different pattern exists here:

https://github.com/tidymodels/yardstick/blob/4a620f3f41289a53e7eda75ae46354c616e35755/R/prob-roc_auc.R#L212-L219

This is because `direction = "minimize"`?

Reply:
It should have been `"maximize"` 😅

Reply:
I'm still a bit confused here, then—if we want to maximize this metric, shouldn't we pass `specificity` rather than `1 - specificity`, as in the code linked above? Or is the thing labeled `specificity` actually `one_minus_specificity` in the first place?